### PR TITLE
refactor!: reduce name conflicts

### PR DESCRIPTION
**Pull request by fubuloubu**
### What I did

This PR makes a breaking change to how the message struct signing info is obtained, which reduces naming conflicts (should eliminate all common ones) when using this package (e.g. #9)

fixes: #9
fixes: #32 (00b4e8c)
fixes: #2 (f704f81)
fixes: #1 (e789713)

### How I did it

Liberal use of sunder properties, some magic

### How to verify it

Check with some common signing flows (I added EIP2612 Permit and Gnosis SafeTx messages)

### Checklist

- [x] Passes all linting checks (pre-commit and CI jobs)
- [x] New test cases have been added and are passing
- [x] Documentation has been updated
- [x] PR title follows [Conventional Commit](https://www.conventionalcommits.org/en/v1.0.0/) standard (will be automatically included in the changelog)

### CODE REVIEW

**Review by unparalleled-js — APPROVED** [review]
LGTM
Some EIP-7127 stuff that worked before continue to work for me after this PR.
Is there any other testing you would like done?

**Comment by fubuloubu** [comment]
> LGTM Some EIP-7127 stuff that worked before continue to work for me after this PR. Is there any other testing you would like done?

Yeah, I think this refactor is looking decently good, although it definitely is a breaking change so I wanted to understand all of the downstream impact before releasing it

Also was trying to refactor into a better dataclass library, but none had the support for the featureset we're using from dataclassy

**Comment by fubuloubu** [comment]
Once this is mergeable, I'd like to update the pin in ape core: https://github.com/ApeWorX/ape/blob/main/setup.py#L122

File: eip712/common.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+        class SafeTx(SafeTxV1):
+            _verifyingContract_ = contract_address
+
+    elif not chain_id:
+        raise ValueError("Must supply 'chain_id=' for Safe versions 1.3.0 or later")
+
+    else:
+
+        class SafeTx(SafeTxV2):  # type: ignore[no-redef]
```
Comment (unparalleled-js) [inline eip712/common.py:107]: Note: You can programmatically define classes using the `type()` fn e.g.

```python
type("SafeTx", (,), {"_verifyingContract_": contract_address})
```

In this situation, both `flake8` and I may prefer.

**Review by unparalleled-js — APPROVED** [review]
Looks good :)

Here is general feedback:

* `if` statements / logic branching in tests should be avoided. Test cases should be narrower and one shouldn't have to question about whether the branches are still relevant.
* I would have expected the tests for `create_safe_tx_def` to be in `test_common.py`. It would be easier to audit this way.

File: eip712/common.py (hunk 1 above)
Comment (fubuloubu) [reply eip712/common.py:107]: I don't know how to this works with how dataclassy abuses Metaclasses for it's functionality, nor how to add class vars w/ annotations and no values

Comment (unparalleled-js) [reply eip712/common.py:107]: Oh, right, the annotations. Perhaps we can re-visit or explore at another time; what you have now is clearly ok!

**Comment by fubuloubu** [comment]
> `if` statements / logic branching in tests should be avoided. Test cases should be narrower and one shouldn't have to question about whether the branches are still relevant.

In this scenario, it's just switching out to a different value, which will never change because the transaction def aligns with an on-chain deployment, and must produce a specific hash

> I would have expected the tests for `create_safe_tx_def` to be in `test_common.py`. It would be easier to audit this way.

Updated this

**Review by unparalleled-js — APPROVED** [review]

**Pull request merged**